src/posix/platform/netif.cpp

```diff
@@ -1232,7 +1232,11 @@ static void processTransmit(otInstance *aInstance)
         message = otIp6NewMessage(aInstance, &settings);
 #endif
         VerifyOrExit(message != nullptr, error = OT_ERROR_NO_BUFS);
+#if OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
+        otMessageSetOrigin(message, OT_MESSAGE_ORIGIN_HOST_TRUSTED);
```

Comment:
I'm fine with changing this when it's used as a reference device. However, it's going to break the openthread CI because the CI is using `OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE=1`.

Reply:
@superwhd I would assume that the CI tests need to be performed on a 'normal' Thread device/stack, not on a reference-device stack. That's because the reference-device flag may modify the 'normal' behavior of a Thread device in special ways, and is also unsecure and non-fit for production use, and so it's not the behavior we would ideally want to test during a CI. Maybe there are just some specific instances where the CI needs a small part of the reference-device functionality?

Reply:
Thanks. I think either way is fine.

Reply:
Some terminology confusion maybe - often (eg in medical device testing), a "reference device" does not have to behave like the DUT necessarily. It may only contain aspects of the DUT behavior, or just be there for reference: e.g. if the Thread DUT needs to change behavior based on measured RF noise level, then the reference device may be just a laboratory-grade RF amplifier with an antenna attached. 

So in OpenThread it wasn't intended like this? If the reference device needs to act the same as a 'normal' Thread device then it makes sense to add extra specific flags for behaviors that are needed for the "test bed devices" (you could then also call these "stimulus devices" - not necessarily Thread compliant but needed to create particular stimuli for the DUT.)   Both the proposed approaches are fine for me.

Reply:
From what I recall, the intent behind REFERENCE_DEVICE was to enable additional logging and extra APIs for testing.

The term "REFERENCE_DEVICE" was introduced to OpenThread in the following PR:
- #3960

Over time in a sequence of PRs, new APIs were added which are available only under `REFERENCE_DEVICE` and intended for testing:
- #5537
- #4759
- #6599
- #7077
- #7182

Reply:
Agree with not changing the behavior of stack by `OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE`. How about we guard all extra functionalities required by reference devices with build time condition and also requires runtime explicit enabling them?

For this particular issue, perhaps we can create a command to send TMF messages. The command would call some new OpenThread API(only available in reference device build and when test mode is enabled) so it will bypass the host network stack(`netif.cpp`). Something like:

```bash
> test enable # enable reference device functionalities
Done
> test tmp send fd00:db9:0:0:0:ff:fe00:ac00 61631 -x 410258088bb161026165ff03087771e1d0000000030010fd007d037d037d03de01c0e036227901
Done
```

Thoughts?

Reply:
@abtink Thanks, yes also the previous name "CERT_LOG" clearly shows that it started by getting extra logging from a device for certification purposes.

@bukepo  If we implement the new command `test tmf send ...` - wouldn't that have the same issue that the filter intercepts the UDP message with dst port = 61631 ?     
And if we apply your solution with a generic prefix CLI command `test` then it would also mean that some other commands that are for reference-devices only need to be integrated into the `test` command, correct?

Maybe just adding a CLI command to disable the filter (on reference-devices only) would be enough for this case.

Reply:
> If we implement the new command test tmf send ... - wouldn't that have the same issue that the filter intercepts the UDP message with dst port = 61631 ?

If we provide API, it won't have the untrusted origin problem. I think the reference device enabled `PLATFORM_UDP` feature, so the socket created by `udp` command creates the socket on the host stack instead of OpenThread stack.

> And if we apply your solution with a generic prefix CLI command test then it would also mean that some other commands that are for reference-devices only need to be integrated into the test command, correct?

Correct.

Reply:
> So basically, a CLI command to toggle the filter for reference devices rather than toggle the message origin in posix netif.

Yes. Sounds good to me.